### PR TITLE
8323724: Remove potential re-inflation from FastHashCode under LM_LIGHTWEIGHT

### PR DESCRIPTION
**Pull request by fbredber**
There is a race when a thread that tries to install the hash code finds an inflated `ObjectMonitor`, which the deflation thread is concurrently deflating, thus causing the `FastHashCode` logic to re-inflate the `ObjectMonitor`.

This PR solves that by simply taking the ObjectMonitor pointer from the markWord and retrying without inflating, if it was deflated.

This enhancement will remove the last code path which can inflate an `ObjectMonitor` from a non java thread, which in turn will simplify the `is_lock_owned` logic and allow for stronger assertions.

Tested and passes tier1-tier5 ok.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323724](https://bugs.openjdk.org/browse/JDK-8323724): Remove potential re-inflation from FastHashCode under LM_LIGHTWEIGHT (**Enhancement** - P4)

### Reviewers
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18876/head:pull/18876` \
`$ git checkout pull/18876`

Update a local copy of the PR: \
`$ git checkout pull/18876` \
`$ git pull https://git.openjdk.org/jdk.git pull/18876/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18876`

View PR using the GUI difftool: \
`$ git pr show -t 18876`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18876.diff">https://git.openjdk.org/jdk/pull/18876.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18876#issuecomment-2067738207)

### CODE REVIEW

**Review by xmas92 — APPROVED** [review]
lgtm.

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 1]
```diff
@@ -1077,9 +1077,11 @@ intptr_t ObjectSynchronizer::FastHashCode(Thread* current, oop obj) {
 
     // Inflate the monitor to set the hash.
 
-    // An async deflation can race after the inflate() call and before we
-    // can update the ObjectMonitor's header with the hash value below.
-    monitor = inflate(current, obj, inflate_cause_hash_code);
+    // There's no need to inflate if the mark has already got a monitor.
+    // NOTE: an async deflation can race after we get the monitor and
+    // before we can update the ObjectMonitor's header with the hash
+    // value below.
```
Comment (coleenp) [inline src/hotspot/share/runtime/synchronizer.cpp:1083]: This is subtle.  The inflate logic will create a new inflated monitor, if the current one is being async deflated, but we can install the hash in that monitor (without it going away? It needs a handshake to go away so no.) and the logic below this comment will handle the async deflation race, which you have even if you call inflate.  Did I get this right?  Rereading the comment implies this is handled below, and in fact it is.

**Review by coleenp — APPROVED** [review]
This looks good.

**Comment by coleenp** [comment]
>This enhancement will remove the last code path which can inflate an ObjectMonitor from a non java thread, which in turn will simplify the is_lock_owned logic and allow for stronger assertions.

Is there an RFE for this?  It looks like a good cleanup.

**Comment by xmas92** [comment]
> > This enhancement will remove the last code path which can inflate an ObjectMonitor from a non java thread, which in turn will simplify the is_lock_owned logic and allow for stronger assertions.
> 
> Is there an RFE for this? It looks like a good cleanup.

`is_lock_owned` was removed in #17626 . But the all the `nullptr` checks in `inflate_impl` can be removed and ` current->is_Java_thread()` can be asserted when `LockingMode == LM_LIGHTWEIGHT`. 

I do not believe there is an RFE for this.

**Comment by coleenp** [comment]
Not necessary to create an RFE before integrating.

**Comment by fbredber** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**